docs/logging.md

```diff
@@ -0,0 +1,73 @@
+# Logging
+
+This code uses the [zap](https://pkg.go.dev/go.uber.org/zap) logging framework.
+
+Zap is a production quality logging library that is performant and highly configurable.
+Zap produces structured logging that can be easily machine-parsed,
+so that people can easily use tooling to filter through logs,
+or logs can easily be sent to log aggregation tooling.
+
+## Conventions
+
+### Instantiation
+
+Structs should hold a reference to their own `*zap.Logger` instance, as an unexported field named `log`.
+Callers should provide the logger instance as the first argument to a `NewFoo` method, setting any extra fields with the `With` method externally.
+For example:
+
+```go
+path := "/tmp/foo"
+foo := NewFoo(log.With(zap.String("path", path)), path)
+```
```

Comment:
I only present this as an alternative perspective. I am not suggesting we make any changes.

This is a polarizing opinion. I like using a global (singelton) logger.

Typically the default singleton is a no-op. 

In `main()` you initialize the singleton once. 

E.g. 

```go
func main() {
    log.Init(out, "json", "debug")
}
```

A factory method lets you grab a logger at any time.

```go
log.Global().Info("Hi!")

aService := SomeService{
    log: log.Global().WithField("service", "SomeService")
}
```

The benefit is simplifying constructors and other callers. You no longer have to toil with passing a logger down a call stack. 

It has the tradeoffs of any singleton and global state, of course. Additionally, you could iadvertently hide logging in a performance critical code path. (But profiling will show it immediately.)

But, I've successfully used this singleton pattern in large-scale, microservice  systems with low toil. The diagnositc quality of the logs was very high.